drivers/pwm/pwm_nrfx.c

```diff
@@ -5,93 +5,258 @@
+	 */
 	const struct pwm_nrfx_config *pconfig = dev->config->config_info;
 	struct pwm_nrfx_data *pdata = dev->driver_data;
+	NRF_PWM_Type *pwm_instance = pconfig->pwm.p_registers;
+	u8_t channel;
+	u32_t ret;
 
-	if (pwm >= NRF_PWM_CHANNEL_COUNT) {
+	/* Check if PWM pin is one of the predefiend DTS config pins.
+	 * Return its array index (channel number),
+	 * or NRF_PWM_CHANNEL_COUNT if not initialized through DTS.
+	 */
+	channel = pwm_channel_map(pconfig, pwm);
+	if (channel == NRF_PWM_CHANNEL_COUNT) {
+		LOG_ERR("PWM pin %d not enabled through DTS configuration.",
+			pwm);
 		return -EINVAL;
 	}
 
-	if (period_cycles != pdata->top_value) {
-		if (period_cycles > PWM_COUNTERTOP_COUNTERTOP_Msk) {
-			return -EINVAL;
-		}
-
-		pdata->top_value = period_cycles;
-		nrf_pwm_configure(pconfig->pwm.p_registers,
-				  pconfig->config.base_clock,
-				  pconfig->config.count_mode,
-				  pdata->top_value);
+	/* Check if period_cycle is either matching currently used, or
+	 * possible to use with our prescaler options.
+	 */
+	ret = pwm_period_check_and_set(pconfig, pdata, pwm_instance, pwm,
+		period_cycles);
+	if (ret) {
+		LOG_ERR("Incompatible period %d", period_cycles);
+		return ret;
 	}
 
-	if (pulse_cycles > pdata->top_value) {
+	/* Check if pulse is bigger than period, fail if so */
+	if (pulse_cycles > period_cycles) {
 		return -EINVAL;
 	}
 
-	/* Modify only the COMPARE bits while preserving the POLARITY
-	 * bit that controls the inversion of the channel.
-	 * For more details see product specification.
+	/* Store new pulse value bit[14:0], and polarity bit[15] for channel. */
+	pdata->current[channel] = (
+		(pdata->current[channel] & NRFX_PWM_POLARITY_MSK)
+		| (pulse_cycles >> pdata->prescaler)
+		);
+
+	/* If Channel is off/fully on (duty 0% or 100%), also set GPIO register
+	 * since this will the setting if we in the future disable the
+	 * peripheral when no channels are active.
 	 */
-	pdata->current[pwm] = ((pdata->current[pwm]
-				& ~PWM_COUNTERTOP_COUNTERTOP_Msk)
-				| pulse_cycles);
+	if (pulse_cycles == 0 || pulse_cycles == period_cycles) {
+		/* If pulse 0% and pin not inverted, set LOW.
+		 * If pulse 100% and pin inverted, set LOW.
+		 * If pulse 0% and pin inverted, set HIGH.
+		 * If pulse 100% and pin not inverted, set HIGH.
+		 */
+		bool channel_inverted_state =
+			get_pin_inversion_state(
+				pconfig->initial_config.output_pins[channel]);
+
+		bool pulse_0_and_not_inverted =
+			(pulse_cycles == 0)
+			&& !channel_inverted_state;
+		bool pulse_100_and_inverted =
+			(pulse_cycles == period_cycles)
+			&& channel_inverted_state;
+
+		if (pulse_0_and_not_inverted || pulse_100_and_inverted) {
+			nrf_gpio_pin_clear(pwm);
+		} else {
+			nrf_gpio_pin_set(pwm);
+		}
+	}
+
+	/* Check if all channels are off (duty 0% or 100%) */
+	if (!any_channel_active(pdata)) {
+		/* If peripheral in use and all channels off: disable. */
+		if (pdata->pwm_instance_in_use) {
+			nrf_pwm_disable(pwm_instance);
```

Comment:
The sequence should be stopped first to avoid any glitch (it depends on the used period length, but it is possible at this point that the new value written to `current` is not effective yet - the old one is still played; and `nrf_pwm_disable` acts immediately).
Actually, I wonder if `nrfx_pwm_stop` wouldn't be enough. Perhaps there is no need to disable the peripheral completely (this will be anyway done by power management when a low power state is entered).

Reply:
As discussed, nrfx_pwm_stop() would be enough.